scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+	}
+
+	// This logic could probably be somewhere else
+	var latestID int
+	if len(keys) == 0 {
+		latestID = 0
+	} else {
+		latestID, err = getLatestId(keys)
+		if err != nil {
+			return TaskMetadata{}, err
+		}
+	}
+
+	metadata := TaskMetadata{
+		ID:          TaskID(latestID + 1),
+		Name:        name,
+		TaskType:    tType,
+		Target:      target,
+		TargetType:  target.Type(),
+		DateCreated: time.Now().UTC(),
+	}
+
+	// I do this serialize and deserialize a lot in this file. Would be nice to have set and get helpers that deal with
+	// all the converting instead
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	key := TaskMetadataKey{taskID: metadata.ID}
+	err = m.storage.Create(key.String(), string(serializedMetadata))
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	runs := TaskRuns{
+		TaskID: metadata.ID,
+		Runs:   []TaskRunSimple{},
+	}
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	taskRunKey := TaskRunKey{taskID: metadata.ID}
+	err = m.storage.Create(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		err.AddDetail("task_id", string(metadata.ID))
+		return TaskMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+func (m *TaskMetadataManager) GetTaskByID(id TaskID) (TaskMetadata, fferr.GRPCError) {
+	key := TaskMetadataKey{taskID: id}.String()
+	metadata, err := m.storage.Get(key)
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	// Should enum 0 as EmptyList or something
+	if len(metadata) == 0 {
+		return TaskMetadata{}, fferr.NewInternalError(fmt.Errorf("task not found for id: %s", string(id)))
+	}
+
+	taskMetadata := TaskMetadata{}
+	err = taskMetadata.Unmarshal([]byte(metadata))
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+	return taskMetadata, nil
+}
+
+func (m *TaskMetadataManager) GetAllTasks() (TaskMetadataList, fferr.GRPCError) {
+	metadata, err := m.storage.List(TaskMetadataKey{}.String())
+	if err != nil {
+		return TaskMetadataList{}, err
+	}
+
+	tml, err := m.getAllTasksAsMetadataList(metadata)
+	if err != nil {
+		return TaskMetadataList{}, err
+	}
+	return tml, nil
+}
+
+func (m *TaskMetadataManager) getAllTasksAsMetadataList(metadata map[string]string) (TaskMetadataList, fferr.GRPCError) {
+	tml := TaskMetadataList{}
+	for _, meta := range metadata {
+		taskMetadata := TaskMetadata{}
+		err := taskMetadata.Unmarshal([]byte(meta))
+		if err != nil {
+			return TaskMetadataList{}, err
+		}
+		tml = append(tml, taskMetadata)
+	}
+	return tml, nil
+}
+
+func (m *TaskMetadataManager) CreateTaskRun(name string, taskID TaskID, trigger Trigger) (TaskRunMetadata, fferr.GRPCError) {
+	// ids will be generated by TM
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task name", name)
+		return TaskRunMetadata{}, err
+	}
+
+	// Not sold on this naming for this struct. Maybe like RunHistory or something?
```

Comment:
I like runhistory but the implication with history is that it's usually sorted by timei